# Laboratorio 2

### Modelos no paramétricos

### Clasificaci&oacute;n con K-Nearest-Neighbors & Ventana de Parzen

### Universidad de Antioquia

### Facultad de Ingeniería

### Ingeniería de Sistemas

### Ude@ - 2018-I

#### Profesor: Antonio Tamayo Herrera


### Integrantes:

#### Nombre: David de Jesus Yepes Isaza
#### Cédula: 1020455542

#### Nombre: Rodrigo Antonio Porras Martinez
#### Cédula: 1039459720



## Guía del laboratorio

En este archivo va a encontrar tanto celdas de código cómo celdas de texto con las instrucciones para desarrollar el laboratorio.

Lea atentamente las instrucciones entregadas en las celdas de texto correspondientes y proceda con la solución de las preguntas planteadas.

Nota: no olvide ir ejecutando las celdas de código de arriba hacia abajo para que no tenga errores de importación de librerías o por falta de definición de variables.

In [21]:
import numpy as np
import matplotlib.pyplot as plt
from __future__ import division
from scipy import stats
from scipy.spatial import distance

#Método para calcular la distancia entre dos vectores
def dist_euclidiana(x1, x2):
    
    dist = distance.euclidean(x1, x2)
    
    return dist


## Ejercicio 1

Analice los siguientes métodos a la luz de la teoría vista para los modelos de K vecinos más cercanos (KNN) y ventana de Parzen. Una vez comprenda su funcionamiento proceda a completar el código del método KNN.


In [70]:
from scipy.stats import mode
import operator

#Error cuadrático medio para los problemas de regresión
def ECM(Y_est,Y):
    N = np.size(Y)
    ecm = sum((Y_est - Y)**2)/(2*N)
    
    return ecm

#Para calcular el error en los problemas de clasificación
def error(Y_lest, Y):
    error = 0
    for ye, y in zip(Y_lest, Y):
        if ye != y:
            error += 1
    
    error = error/np.size(Y)
    
    return error


def KNN(X_train, Y_train, X_val, Y_val, knn, tipo):
    
    k = knn    #Parámetro k que equivale al número de vecinos a tener en cuenta para resolver el problema de
               #predicción de la variable de salida
    
    Y_est = []
    
    for x1 in X_val:
        distances = dict()
        for x2, y2 in zip(X_train, Y_train):
            distances.update({dist_euclidiana(x1,x2): y2})
    
        #k vecinos más cercanos cercanos
        vecinos = sorted(distances.items(), key=operator.itemgetter(0))
        #print vecinos
        
        #Complete aquí el código para tomar solo la variables de salida de los k vecinos más cercanos
        y_vecinos = []
        for v in vecinos[:knn]:
            y_vecinos.append(v[1])
        
        #Se calcula la variable de salida
        #Regression
        if tipo == 'regression':
            #Complete el código aquí para calcular la variable de salida en el problema de regresión
            #y_est = 
            Y_est.append(y_est)
        #Classification
        elif tipo == 'classification':
            #Complete el código aquí para calcular la variable de salida en el problema de clasificación
            mode = stats.mode(y_vecinos, axis=None)
            #print(mode)
            #print('mode[0]:'+mode[0])
            y_est = mode[0]
            Y_est.append(y_est[0])
            #print(Y_est)
    
    if tipo == 'regression':
        print ("Modelo KNN para regresión\n\n"+"El Error cuadrático medio del modelo con " + str(k) + " vecinos es = " + str(ECM(Y_est, Y_val)))
    elif tipo == 'classification':
        print ("Modelo KNN para clasificación\n\n"+"El Error de clasificación del modelo con " + str(k) + " vecinos es = " + str(error(Y_est,Y_val)))
        

## Ejercicio 2

Ahora va a resolver un problema de clasificación con el método KNN. A continuación se leen los datos de un problema de clasificación triclase. Las variables o caracterísicas son guardadas en la variable X que posteriormente es normalizada y la variable de salida (clase $0$, $1$ o $2$ ) es guardada en la variable Y. Note que la base de datos de entrenamiento se parte, de manera aleatoria, para entrenamiento y validación. Analice el siguiente c&oacute;digo y responda a las siguientes preguntas.

In [92]:
#cargamos la bd iris desde el dataset de sklearn
from sklearn import datasets
from sklearn.preprocessing import normalize

iris = datasets.load_iris()

#Esta es la base de datos Iris del UCI Machine Learning Repository 
#https://archive.ics.uci.edu/ml/datasets/iris

X, Y = iris.data, iris.target


#Normalizamos los datos
X = normalize(X, axis=0, norm='l1')

#Partición de la base de datos
Nsample = np.size(X,0);
#print(Nsample)
indices = np.random.permutation(Nsample);
#print(np.shape(indices))
#print(indices)
#print(int(round(0.7*Nsample)))

#Conjunto de muestras de entrenamiento
Xtrain = X[indices[1:int(round(0.7*Nsample))],:]
Ytrain = Y[indices[1:int(round(0.7*Nsample))]]

#Conjunto de muestras de validación
Xtest = X[indices[int(round(0.7*Nsample))+1:Nsample],:]
Ytest = Y[indices[int(round(0.7*Nsample))+1:Nsample]]

2.1 Describa la base de datos de entrenamiento. Diga cu&aacute;ntas muestras tiene en total y cu&aacute;ntas caracter&iacute;sticas.

2.2 Qu&eacute; porcentaje de la base de datos se usa para el entrenamiento del modelo? A cu&aacute;ntas muestras corresponde esto? Cu&aacute;ntas muestras se usan para la validaci&oacute;n?

Responda Aqu&iacute;:

2.1 R/: El conjunto de datos contiene $3$ clases de $50$ instancias cada una, donde cada clase se refiere a un tipo de planta de iris. Una clase es linealmente separable de los otros $2$; estos últimos NO son linealmente separables el uno del otro. Tiene $150$ muestras, $50$ por cada clase de flor y tiene $4$ caracteristicas todas son numericas y en Centimestros: largo y ancho del sepalo y largo y ancho del petalo

2.2 R/: Para el entrenamiento se usa el $70\%$ y corresponde a  $105$ muestras y para la validacion se usan $45$ muestras 

In [94]:
##### Completa el código llamado al método KNN con sus respectivos argumentos para el problema de clasificación
tipo="classification"
for k in range(1,12,2):
    KNN(Xtrain,Ytrain,Xtest,Ytest,k,tipo)

Modelo KNN para clasificación

El Error de clasificación del modelo con 1 vecinos es = 0.06818181818181818
Modelo KNN para clasificación

El Error de clasificación del modelo con 3 vecinos es = 0.06818181818181818
Modelo KNN para clasificación

El Error de clasificación del modelo con 5 vecinos es = 0.06818181818181818
Modelo KNN para clasificación

El Error de clasificación del modelo con 7 vecinos es = 0.06818181818181818
Modelo KNN para clasificación

El Error de clasificación del modelo con 9 vecinos es = 0.06818181818181818
Modelo KNN para clasificación

El Error de clasificación del modelo con 11 vecinos es = 0.06818181818181818


## Ejercicio 3

Identifique la variable k (número de vecinos más cercanos definido para el modelo KNN) en el método KNN. En la celda de c&oacute;digo anterior, llame al método KNN cambiando el valor de $k$ de acuerdo a la siguiente tabla y complete la columna Error de clasificaci&oacute;n.

<table>
  <tr>
    <th>K (número de vecinos)</th>
    <th>Error de clasificación</th>
  </tr>
  <tr>
    <td>1</td>
    <td>0.09090909090909091</td>
  </tr>
  <tr>
    <td>3</td>
    <td>0.06818181818181818</td>
  </tr>
  <tr>
    <td>5</td>
    <td>0.06818181818181818</td>
  </tr>
  <tr>
    <td>7</td>
    <td>0.06818181818181818</td>
  </tr>
  <tr>
    <td>9</td>
    <td>0.06818181818181818</td>
  </tr>
  <tr>
    <td>11</td>
    <td>0.06818181818181818</td>
  </tr>
    
</table>


3.1 En el modelo KNN, qué criterio estadístico está usando para tomar la decisión final y asignar la clase a una nueva muestra y por qué? 


Responda aquí:
    
3.1 R/: Se toma el criterio de la moda y de preferencia que K sea impar,  des esta manera seleciona la clase de los vecinos que estan mas cerca. 

3.2 ¿Por qué cree que se deben armar los conjuntos de entrenamiento y
prueba de forma aleatoria? Analice la forma como se carga la base de datos y a partir de all&iacute; argumente su respuesta.

Responda aquí:
    
3.2 R/: Al formar conjuntos de muestras de maneras aleatorias, evitamos que el modelo solo sea efectivo para un grupo de validacion especifico. Comparamos los resultados que tiene con cada grupo de muestras y podemos notar que el error no es constante pero esta entre un intervalo de 3 a 1 muestras mal predichas por grupo de 45 muestras de validacion. Existe ocasiones en los que  el error de clasificacion es 0, pero al ejecutar varias veces vemos que solo pasa para ese grupo aleatorio de muestras. Para la reconocer que el modelo esta predicciendo correctamente las clases se debe probar con muchas grupos de muestras.

## Ejercicio 4

A continuación debe completar el código del método parzen para resolver problemas de clasificación con el modelo ventana de Parzen.

In [171]:
#Calcula el kernel gaussiano de x
def kernel_gaussiano(x):
    return (1/np.sqrt(2*np.pi))*np.exp((-0.5)*x**2)

def parzen(X_train1, Y_train1, X_train2, Y_train2, X_train3, Y_train3, X_val, Y_val, ancho_h):
    
    h = ancho_h
    d=np.size(X_val,1)
    #print(d)
    model1 = []
    model2 = []
    model3 = []
    Y_est = []
    
    for x1 in X_val:
        
        y_est1 = 0
        distances1 = []
        
        y_est2 = 0
        distances2 = []
        
        y_est3 = 0
        distances3 = []
        
        for x2_1, x2_2, x2_3 in zip(X_train1, X_train2, X_train3):
            #Complete aquí el código para calcular las ditancias en los tres modelos
            distances1.append(dist_euclidiana(x1,x2_1))
            distances2.append(dist_euclidiana(x1,x2_2))
            distances3.append(dist_euclidiana(x1,x2_3))
    
        #Complete el código aquí con las fdp para los tres modelos
        #implemente los modelos de ventana de Parzen por clase
        #distances1=np.asarray(distances1)
        #distances2=np.asarray(distances2)
        #distances3=np.asarray(distances3)
        y_est1 = sum(kernel_gaussiano(np.divide(distances1,h)))/(np.size(X_train1,0)*(h**d))  
        model1.append(y_est1)
        
        y_est2 =  (sum(kernel_gaussiano(np.divide(distances2,h))))/(np.size(X_train2,0)*(h**d))
        model2.append(y_est2)
        
        y_est3 =  (sum(kernel_gaussiano(np.divide(distances3,h))))/(np.size(X_train3,0)*(h**d))
        model3.append(y_est3)
   
    #Se asigna la clase a la nueva muestra
    for y1, y2, y3 in zip(model1, model2, model3):
        clase = 0
        if y2 > y1:
            clase = 1
            if y3 > y2:
                clase = 2
        elif y3 > y1:
            clase = 2
    
        Y_est.append(clase)
            
    print ("Modelo ventana de Parzen para clasificación\n\n"+"El Error de clasificación con h = " + str(h) + " es = " + str(error(Y_est, Y_val)))


## Ejercicio 5

Ahora debe resolver un problema de clasificación con el método ventana de Parzen. A continuación se leen los datos de un problema de clasificación triclase. Las variables o caracterísicas son guardadas en la variable X que posteriormente es normalizada y la variable de salida (clase $0$, $1$ o $2$ ) es guardada en la variable Y. Usted debe realizar la separación de las observaciones muestrales por clase, recuerde que este es un requisito para resolver problemas de clasificación con este modelo y que debe implementar un modelo por cada clase del problema.
Una vez tenga la base de datos de entrenamiento separada por clases, haga una partición de manera aleatoria en un 70% de las muestras para entrenamiento y un 30% de las muestras para validación.

In [172]:
#cargamos la bd iris desde el dataset de sklearn
from sklearn import datasets
iris = datasets.load_iris()

#Esta es la base de datos Iris del UCI Machine Learning Repository 
#https://archive.ics.uci.edu/ml/datasets/iris

X, Y = iris.data, iris.target


#Normalizamos los datos
X = normalize(X, axis=0, norm='l1')

#Separación de muestras por clase
#Complete el código aquí (Analice la base de datos de entrenamiento y observe que es muy sencillo hacer la separación)


X0 = X[:50,:]
Y0 = Y[:50]

X1 =X[50:100,:] 
Y1 = Y[50:100]

X2 = X[100:,:]
Y2 = Y[100:]

#print(np.shape(Y2))

#Partición de la base de datos en 70%-30% para entrenamiento y validación respectivamente
Nsample0 = np.size(X0,0);
indices0 = np.random.permutation(Nsample0);

Nsample1 = np.size(X1,0);
indices1 = np.random.permutation(Nsample1);

Nsample2 = np.size(X2,0);
indices2 = np.random.permutation(Nsample2);


#Conjunto de muestras de entrenamiento clase 0
Xtrain0 = X0[indices0[0:int(round(0.7*Nsample0))],:]
Ytrain0 = Y0[indices0[0:int(round(0.7*Nsample0))]]

#Conjunto de muestras de validación clase 0
Xtest0 = X0[indices0[int(round(0.7*Nsample0))+1:Nsample0],:]
Ytest0 = Y0[indices0[int(round(0.7*Nsample0))+1:Nsample0]]

#Conjunto de muestras de entrenamiento clase 1
Xtrain1 = X1[indices1[0:int(round(0.7*Nsample1))],:]
Ytrain1 = Y1[indices1[0:int(round(0.7*Nsample1))]]

#Conjunto de muestras de validación clase 1
Xtest1 = X1[indices1[int(round(0.7*Nsample1))+1:Nsample1],:]
Ytest1 = Y1[indices1[int(round(0.7*Nsample1))+1:Nsample1]]

#Conjunto de muestras de entrenamiento clase 2
Xtrain2 = X2[indices2[0:int(round(0.7*Nsample2))],:]
Ytrain2 = Y2[indices2[0:int(round(0.7*Nsample2))]]

#Conjunto de muestras de validación clase 2
Xtest2 = X2[indices2[int(round(0.7*Nsample2))+1:Nsample2],:]
Ytest2 = Y2[indices2[int(round(0.7*Nsample2))+1:Nsample2]]

#Concatenamos las muestras de validación
Xtest = np.concatenate((Xtest0,Xtest1,Xtest2), axis=0)
Ytest = np.concatenate((Ytest0,Ytest1,Ytest2))
#print(np.size(Ytest))


In [177]:
#Completa la siguiente línea con el llamado al método parzen con sus respectivos argumentos
h=10
parzen(Xtrain0, Ytrain0, Xtrain1, Ytrain1, Xtrain2, Ytrain2, Xtest, Ytest, h)

4
Modelo ventana de Parzen para clasificación

El Error de clasificación con h = 10 es = 0.09523809523809523


## Ejercicio 6

Identifique la variable h (ancho de la ventana $h$) en el método parzen. Llame al método parzen cambiando el valor de $h$ de acuerdo a la siguiente tabla y complete la columna Error de clasificación.

<table>
  <tr>
    <th>h (ancho de la ventana)</th>
    <th>Error de clasificación</th>
  </tr>
  <tr>
    <td>0.001</td>
    <td>0.047619047619047616</td>
  </tr> 
  <tr>
    <td>0.01</td>
    <td>0.07142857142857142</td>
  </tr>
  <tr>
    <td>0.1</td>
    <td>0.07142857142857142</td>
  </tr>
  <tr>
    <td>1</td>
    <td>0.07142857142857142</td>
  </tr>
  <tr>
    <td>10</td>
    <td>0.07142857142857142</td>
  </tr>
</table>


6.1 Por qué el modelo de ventana de Parzen es un modelo no paramétrico? 


Responda aquí:
    
6.1 R/: Es un modelo no parametrico porque no supone el orden o forma de como estan distribuidos los datos